### ISSUE TITLE
shmall not set up correctly

### ISSUE THREAD

**Issue by dalehenrich** [hpi-swa/smalltalkCI#68]
According the [logging from this travis build](https://travis-ci.org/dalehenrich/smalltalkCI/jobs/111380537#L58), the `hostname` is `Traviss-Mac-441.local`, but that name does not show up in /etc/hosts (as required by GemStone):

```
>>hostname
Traviss-Mac-441.local
>> cat /etc/hosts
##
# Host Database
#
# localhost is used to configure the loopback interface
# when the system is booting.  Do not change this entry.
##
127.0.0.1   localhost 
255.255.255.255 broadcasthost
::1             localhost  
fe80::1%lo0 localhost
```

You can look at [this commit](https://github.com/dalehenrich/smalltalkCI/commit/fdf1dea1bb0fd406695df96ad473efef778e38ca) for info about the logging that I added to `gemstone/run.sh`...

**Comment by fniephaus** [comment]
[This](https://github.com/travis-ci/travis-build/blob/0b3a45524eac7a3318fa2ecb6fe5b67ab0064800/lib/travis/build/script/smalltalk.rb#L120) needs to be changed.

**Comment by fniephaus** [comment]
This should now be resolved, but OS X builds are still failing:
https://travis-ci.org/hpi-swa/smalltalkCI/jobs/111893995#L1059

**Comment by dalehenrich** [comment]
Unfortunately I removed the logging that provided the information about what is going wrong so I'll have to add the logging information back in to see what the problem is ...

**Comment by fniephaus** [comment]
Can't you just restart an older build?
Also, I think it'd make sense to keep displaying this debug information in debug mode (`--debug`).
You can just do it like [this](https://github.com/hpi-swa/smalltalkCI/blob/bddb789cacc520ce921f08ddcf17baed690f8028/run.sh#L256).

**Comment by dalehenrich** [comment]
I'll see what happens when I restart an [old build](https://travis-ci.org/dalehenrich/smalltalkCI/jobs/111380537).

Logging info from original build:

```
> hostname
Traviss-Mac-441.local
> cat /etc/hosts
##
# Host Database
#
# localhost is used to configure the loopback interface
# when the system is booting.  Do not change this entry.
##
127.0.0.1	localhost 
255.255.255.255	broadcasthost
::1             localhost  
fe80::1%lo0	localhost

......

Creating stone travis
/bin/hostname: 'Traviss-Mac-379.local
'
NetNameResolver localHostName: '10.182.11.61'
```

**Comment by dalehenrich** [comment]
Looking at this original debug information, I notice that `hostname` is getting two answers in same run:
- `Traviss-Mac-441.local` at very start
- `Traviss-Mac-379.local` later on
- error out with `localhost6` which is **not** `hostname` and **not** in `/etc/hosts`

**Comment by dalehenrich** [comment]
With the new run it looks like the hostname has **not** been added to `/etc/hosts`:

```
> hostname
Traviss-Mac-441.local
> cat /etc/hosts
##
# Host Database
#
# localhost is used to configure the loopback interface
# when the system is booting.  Do not change this entry.
##
127.0.0.1	localhost 
255.255.255.255	broadcasthost
::1             localhost  
fe80::1%lo0	localhost
```

So either the new smalltalk.rb was not picking up by the old build or the new smalltalk.rb is not correctly setting an entry in `/stc/hosts` ...

What do you think? Add the logging, because restarting the old build was a waste of time, or do you have enough information to really fix smalltalk.rb?

**Comment by dalehenrich** [comment]
Oh and exact same [error host on failure](https://travis-ci.org/dalehenrich/smalltalkCI/jobs/111380537#L1095) --- `localhost6`

**Comment by dalehenrich** [comment]
...and I can't even guess where localhost6 is coming from at this point in time ... I think that OS X is different than linux and that the `addHost` trick used in legacy build may have involved additional magic (hidden in a .rb file somewhere)...

**Comment by dalehenrich** [comment]
For the legacy builds the [`set hostname to travis-dev` and `add travis-dev to /etc/hosts`](https://github.com/GsDevKit/GsDevKit_home/blob/master/.travis.yml#L26-L31) worked on both platforms and the [mac builds are functional](https://travis-ci.org/GsDevKit/GsDevKit_home/jobs/111901162#L823-L1333) and produced the following information:

```
0.00s$ hostname
travis.dev
The command "hostname" exited with 0.
0.00s$ cat /etc/hosts
##
# Host Database
#
# localhost is used to configure the loopback interface
# when the system is booting.  Do not change this entry.
##
127.0.0.1	localhost travis.dev
255.255.255.255	broadcasthost
::1             localhost  travis.dev
fe80::1%lo0	localhost
```

Of course I don't know what other places to look on the mac, since I never had to dig deeper ...

**Comment by fniephaus** [comment]
This is odd, because the `smalltalk.rb` should be [doing exactly the same](https://github.com/travis-ci/travis-build/blob/1dad4573964bab27b2e18cb1d4b47b035efaaf6b/lib/travis/build/script/smalltalk.rb#L116) that the [host addon would do](https://github.com/travis-ci/travis-build/blob/1dad4573964bab27b2e18cb1d4b47b035efaaf6b/lib/travis/build/addons/hosts.rb#L18).

Maybe `$(hostname)` returns nothing on Travis? I will check that, but it'd need to wait until sometime next week.

**Comment by fniephaus** [comment]
Also, we could try to enable the addon to see if that fixes the problem.

**Comment by dalehenrich** [comment]
How do you enable the addon when you don't know the name of the host, 
can you define an env var in bash that is available for use in the addon ?

**Comment by fniephaus** [comment]
An env var doesn't work, I think.
But we could hardcode travis.dev for OS X builds.
Since those builds don't run in a container, we should be able to change the hostname.

**Comment by dalehenrich** [comment]
how do you arrange to run a sudo command only for OSX, since hardcoding travis.dev also involves changing the hostname of the vm?

**Comment by fniephaus** [comment]
Check if it's a OS X build and then run the `sudo` command?
`sudo` is enabled in OS X builds:
https://docs.travis-ci.com/user/osx-ci-environment/#Overview

But we can just make this part of the `smalltalk.rb` when it works.

`sudo` is only disabled for containers, that's because Docker ;)

**Comment by dalehenrich** [comment]
how do you "check if it's a osx build"? I ask these questions because I am not familiar with yaml ... there seem to be dozens of ways to do each thing, but only a half dozen seem to work in travis ... I know I am probably missing a key concept, but I'm tired of randomly changing the yaml file until travis does what i want:), so I asl questions and am looking for more specific  answers ...

**Comment by fniephaus** [comment]
Well I guess you could use the host addon within the build matrix, but you could also just forget about the yaml file and do everything in the bash script. That'd be much easier.

**Comment by dalehenrich** [comment]
you mean hard code things in gemstone/run.sh?

I guess I already know that that will work as it's been proven with the 
earlier legacy builds ... so I can wait until you patch the smalltalk.rb 
file since that is the ultimate solution, right?

I have yet to have a failure on the mac that doesn't also appear on 
linux, so it's not critical for me...

**Comment by fniephaus** [comment]
@dalehenrich I'm getting an argument error when hardcoding the hostname:
https://travis-ci.org/hpi-swa/smalltalkCI/jobs/112704054#L1082
Maybe the script doesn't like OS X instances that are IPv6-connected?

**Comment by dalehenrich** [comment]
well, you `echo $hostname` isn't [dumping anything](https://github.com/hpi-swa/smalltalkCI/blob/d4a467472c0e5b9bb9d7def700355d7a65d77ded/gemstone/run.sh#L111) so we are not learning whether or not you are successfully setting the hostname ... and the `/etc/hosts/`doesn't like it had anything added to it .,..

With that said, there is an indication that the hostname _is_ travis.dev:

```
Creating stone travis
/bin/hostname: 'travis.dev
'
NetNameResolver localHostName: '127.0.53.53'
	Creating product symbolic link to GemStone64Bit3.3.0-i386.Darwin
	Creating subdirectories 
	Creating stats subdirectories 
	Copying system.conf file
	Copying template files
	Create .topazini file
	Creating defStone.env symbolic link to /Users/travis/build/hpi-swa/smalltalkCI/_builds/GsDevKit_home/bin/defStone.env
	Finished creating stone travis
```
but without the hostname in the `/etc/hosts/` fiel it isn't likely to work ...

And of course, I don't know where the name `localhost6` is coming from ... 

```
Network lookup failure; could not find server 'travis' on host 'travis.dev' because file not found: /opt/gemstone/locks/travis..LCK;  service travis not found ; NetLDI service 'gs64ldi' not found on node 'localhost6' port 50377 :
connect(5,::ffff:127.0.0.1,port=50377) failed with errno=22,EINVAL, Invalid argument (programmer error)
```

**Comment by fniephaus** [comment]
Yes, I realized that `echo $hostname` doesn't make any sense :sweat_smile: 
Nonetheless, I can't explain `localhost6`. Are you sure that the script is working on OS X with IPv6 enabled?

**Comment by dalehenrich** [comment]
[This build ran on osx](https://github.com/dalehenrich/smalltalkCI/blob/57d1dc180d0ba375dfb9b1108346e19cc3168980/.travis.yml) ... but I believe that this was with sudo enabled ...

Using [`addons - hosts travis.dev`](https://github.com/dalehenrich/smalltalkCI/blob/57d1dc180d0ba375dfb9b1108346e19cc3168980/.travis.yml#L29-L31) and[ `sudo hostname travis.dev`](https://github.com/dalehenrich/smalltalkCI/blob/57d1dc180d0ba375dfb9b1108346e19cc3168980/gemstone/run.sh#L114) ... now I see that you are trying to use `sudo scutil --set HostName travis.dev` which I haven't tried ...

GemStone should work with ipv6 ... perhaps scutil isn't quite doing the right thing as far as GemStone is concerned ...

**Comment by dalehenrich** [comment]
I'm pretty sure that OSX has multiple hostname sites and the perhaps the `scutil` call changes the hostname that is visible to the user, but not the hostname that GemStone needs ... I'm pretty sure that I ran into this problem (scutil not doing the trick) when I was getting my mac laptop setup and that's why I've been using `sudo hostname` ...

**Comment by fniephaus** [comment]
Mh interesting, this build seemed to be [almost successful](https://travis-ci.org/dalehenrich/smalltalkCI/jobs/100459838#L1456).
I will start another test build...

**Comment by dalehenrich** [comment]
Yeas, but to get to the point where the API rate limit gets hit, you have to be logged into a stone ... so it's getting past the error we are seeing ...

**Comment by dalehenrich** [comment]
BTW, I think I've [fixed the bug in Metacello that was leading to the API rate limit issues](https://github.com/dalehenrich/metacello-work/issues/385) ... Given the random nature of the failures, it will take a while to know for sure ... OSX was pretty consistent in hitting the API rate limit, which is why it's been on the allow failures list everywhere I turn it on:)

**Comment by fniephaus** [comment]
Maybe `localhost6` is because I'm also adding the hostname to the IPv6 address in `/etc/hosts`.
But I should've done exactly the [same as the hosts addon](https://github.com/travis-ci/travis-build/blob/dcabdfe556a955cbab15a6dca78167d9dda8c279/lib/travis/build/addons/hosts.rb#L18). Weird!

Good job fixing that API rate limit issue! :)

**Comment by dalehenrich** [comment]
My guess is that `localhost6` is in `/etc/hostname` or where ever the \<blah\> in `sudo hostname <blah>`  goes

**Comment by fniephaus** [comment]
Let's find out: [334.1](https://travis-ci.org/hpi-swa/smalltalkCI/jobs/113000301) & [335](https://travis-ci.org/hpi-swa/smalltalkCI/builds/113001283).

**Comment by dalehenrich** [comment]
[this is the reference that I used to set my laptop hostname](http://osxdaily.com/2010/09/06/change-your-mac-hostname-via-terminal/) and if I recall correctly the visible hostname change (in the GUI) after scutil command was run, but I had to do `sudo hostname` before GemStone worked ...

**Comment by dalehenrich** [comment]
well  [334.1](https://travis-ci.org/hpi-swa/smalltalkCI/jobs/113000301) was doomed because hostname wasn't set anywhere both `addons -hosts:` and `sudo hostname` are needed...

**Comment by fniephaus** [comment]
both failed, so let's try [336](https://travis-ci.org/hpi-swa/smalltalkCI/builds/113004074).

**Comment by dalehenrich** [comment]
Okay ... [same result and same error](https://travis-ci.org/hpi-swa/smalltalkCI/builds/113001283), but now  we know that the hast names should be setup correctly ... I will now have to add additional logging to GsDevKit_home (on the dev branch) to checkout a few more things

**Comment by dalehenrich** [comment]
I just noticed a point of confusion (for me) ... this error is to be expected:

```
waitstone[Error]: Network service !#dir:/Users/travis/build/hpi-swa/smalltalkCI/_builds/GsDevKit_home/server/stones/travis/logs#log:%N%P.log#server!travis was not found.
Network lookup failure; could not find server 'travis' on host 'travis.dev' because file not found: /opt/gemstone/locks/travis..LCK;  service travis not found ; NetLDI service 'gs64ldi' not found on node 'localhost6' port 50377 :
connect(5,::ffff:127.0.0.1,port=50377) failed with errno=22,EINVAL, Invalid argument (programmer error)
```

It's coming from the `waitstone` command and complains about `/opt/gemstone/locks/travis..LCK` not found ... this is okay because the stone is not running yet ...




this error is the troublesome error:

```Stone was unable to attach to shared cache on localhost
Reason: fork of cache monitor failed, Monitor process (985) did not start.
; NetLDI service ''gs64ldi'' not found on node ''localhost6'' port 50377 :
connect(10,::ffff:127.0.0.1,port=50377) failed with errno=22,EINVAL, Invalid argument (programmer error)
Help: 
  Operating system kernel configured for shared memory?
  Check SharedPageCache Monitor log file (*pcmon.log).
  Check SHR_PAGE_CACHE_NUM_PROCS in stone config file.
  Stale files in /opt/gemstone/locks/<stonename>*?
  ipcs - check for stale resources ?
An error occurred when initializing the shared cache
    Stone startup has failed.
```

and while there is the message about netldi service not found, that may be misleading ... as there are other possible explanations for the failure..

Do you happen to be setting up shared memory correctly on the mac at this point?

**Comment by dalehenrich** [comment]
`kern.sysv.shmmax: 4194304 -> 2147483648` is not the only parameter that needs to be set and of course I have no idea what you are doing because nothing is logged for the GemStone build and I don't know whether you are logging everything or not ...

[kern.sysv.shmmax](https://github.com/GsDevKit/GsDevKit_home/blob/master/bin/utils/installOsPrereqs#L231) needs to be set as well as [kern.sysv.shmall](https://github.com/GsDevKit/GsDevKit_home/blob/master/bin/utils/installOsPrereqs#L243)

**Comment by fniephaus** [comment]
That could be the problem, let's try [337](https://travis-ci.org/hpi-swa/smalltalkCI/builds/113008588). (OS X builds always have sudo enabled)

**Comment by dalehenrich** [comment]
Looks like you were missing the setup of [kern.sysv.shmall: 1024 -> 524288](https://travis-ci.org/hpi-swa/smalltalkCI/builds/113008588#L304-L310)

```
[Info] Setting up shared memory
  Total memory available is 4096 MB
  Max shared memory segment size is 2048 MB
  Max shared memory allowed is 4 MB
[Info] No need to increase max shared memory segment size
[Info] Increasing max shared memory allowed to 2048 MB
kern.sysv.shmall: 1024 -> 524288
[Info] The following shared memory settings already exist in /etc/sysctl.conf
```

Are you missing shmall for Linux as well?

**Comment by dalehenrich** [comment]
Cross fingers that we don't hit API rate limit :)

**Comment by fniephaus** [comment]
[It looks like](https://travis-ci.org/hpi-swa/smalltalkCI/builds/113008588#L310) you're right.
Question: do you want OS X to be set up via `smalltalk.rb` or would you prefer to run `installOsPrereqs` instead? I could just remove the prereq logic from `smalltalk.rb` for OS X and we can enable [this line](https://github.com/hpi-swa/smalltalkCI/blob/3c321baa22a0db2b4494aafbeea06cf7e9e24469/gemstone/run.sh#L28) for Linux builds only.
What do you think?

**Comment by dalehenrich** [comment]
I suppose it would be cleaner to do all of the `sudo` work in smalltalk.rb to avoid having to do any sudo operation in run.sh for GemStone, since thats where all of the setup code is supposed to be located. .. but if you prefer to put have that in run.sh then I won't argue with you... 

As I've said before, I'm not in a hurry to get the mac builds running, so a wait is not a big deal for me ...

**Comment by fniephaus** [comment]
It's totally up to you, really. I just need to figure out why [this](https://github.com/travis-ci/travis-build/blob/dcabdfe556a955cbab15a6dca78167d9dda8c279/lib/travis/build/script/smalltalk.rb#L159) is not working as intended.

On a different note, is there any way to speed up GemStone builds? 95% of the time there seems to be some generic preparation going on. Wouldn't it be way more efficient to use a prepare test image instead?

**Comment by fniephaus** [comment]
And again, you are right. I messed up [here](https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/script/smalltalk.rb#L137) and [here](https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/script/smalltalk.rb#L170).
I guess I will need to talk to my friends at Travis again...but probably later this week

**Comment by dalehenrich** [comment]
Then I prefer to have it smalltalk.rb ...

There are a couple of things that I can do to speed things up ... this work would have to be done in tandem with chnages wot GsDevKit_home, so being able to a) switch branches is critical and b) would be useful to be able to test GsDevKit_home with smalltalkCI before tackling the _cache issues. Also right now I don't have a lot of time play with optimizations ... I'll be happy when everything is just working :) 

Along these lines, how does one arrange to flush the _cache dir? Is the _cache a per user _cache or is it cached for all smalltalkCI builds ... do you intend to have an env var or something that can be used to flush the cache  (like when a new Squeak/Pharo vm is available)...

**Comment by fniephaus** [comment]
No worries, happy to do this in tandem and as soon as you have enough time.
I haven't really thought the whole caching thing through. Currently, `_cache` needs to be flushed manually by the user and this is just an optimization for running smalltalkCI locally. Also, I don't think Travis builds would benefit from "caching" content that can be downloaded, because the cache is also just downloaded. It would make sense to cache prepared images or anything like that. However, the user needs to explicitly enable caching via `.travis.yml` and that also only works for Linux container builds. So, maybe Travis caching isn't what we want after all.

I hope everything makes sense.

**Comment by fniephaus** [comment]
So [this](https://travis-ci.org/hpi-swa/smalltalkCI/builds/113019957) happened last night.
Shall I just fix the shmall problem in `smalltalk.rb` or do you want me to set the hostname to `travis.dev

**Comment by fniephaus** [comment]
So [this](https://travis-ci.org/hpi-swa/smalltalkCI/builds/113019957) happened last night.
Shall I just fix the shmall problem in the `smalltalk.rb` or do you want the hostname to be set to `travis.dev` in OS X as well?

**Comment by dalehenrich** [comment]
I think that all of the sudo/os/hostname mangling should be done in smalltalk.rb ... if anything changes it will have to be changed for both linux and osx and the linux stuff _has_ to be done in smalltalk.rb, so I don't see an advantage to splitting it up ...

**Comment by fniephaus** [comment]
Ok, let me rephrase this: since everything seems to work after fixing the shmall problem, do you think it's still necessary to do a `sudo hostname "travis.dev"`. It's currently working completely without setting the hostname explicitly.

**Comment by dalehenrich** [comment]
If it's not needed, then it doesn't need to be done ... there is no 
where in the GsDevKit_home space or travis space where I expect a 
certain hostname ... At one point I had a bug, where the pharo code was 
truncating the host name, but I've patched that ...

**Comment by fniephaus** [comment]
ok cool! I know what to do then :) thanks!

**Comment by dalehenrich** [comment]
Of course, this implies that `sudo hostname <xxx>` is not needed for `any` value of `<xxx>` .... I think that osx is quirky and I think that travis is quirky ... if travis sets up osx **without** ensuring that  `sudo hostname <xxx>` has been executed with whatever hostname that  they have chosen, then  `sudo hostname <xxx>` is still required (on osx), but we can use the hostname that they have chosen instead of one of our own choosing ...

**Comment by dalehenrich** [comment]
... if that was actually the experiment you tried last night then `sudo hostname <xxx>` does not need to be added for my sake :)

**Comment by fniephaus** [comment]
For some reason, https://github.com/travis-ci/travis-build/pull/662 does not seem to have reached production yet. I will close this once it has and after I removed [the temporary fix](https://github.com/hpi-swa/smalltalkCI/blob/5dc687a0a3bf99e05e5744cb4442faef359d1eff/gemstone/run.sh#L105-L110).

**Comment by fniephaus** [comment]
https://github.com/travis-ci/travis-build/pull/692 should finally closed this issue.
There was (yet another) but in the `smalltalk.rb`.

**Comment by dalehenrich** [comment]
That means I should be able to remove the patch? If so assign to me and I will remote the patch ... Not going to take the time to test right now:)

**Comment by fniephaus** [comment]
Yes, when https://github.com/travis-ci/travis-build/pull/692 is in production. I can keep track of that and remove the patch when the time has come...

**Comment by dalehenrich** [comment]
Ah, cool, so we are still waiting:)
